File: language-module-swift/src/main/kotlin/com/voxeldev/mcodegen/dsl/language/swift/souce/generate/extensions/ConvertFields.kt
package com.voxeldev.mcodegen.dsl.language.swift.souce.generate.extensions

import com.voxeldev.mcodegen.dsl.ir.IrField
import com.voxeldev.mcodegen.dsl.ir.IrVisibilityInternal
import com.voxeldev.mcodegen.dsl.ir.IrVisibilityPrivate
import com.voxeldev.mcodegen.dsl.ir.IrVisibilityPublic
import com.voxeldev.mcodegen.dsl.language.swift.SwiftModule
import com.voxeldev.mcodegen.dsl.language.swift.ir.IrVisibilityFileprivate
import com.voxeldev.mcodegen.dsl.language.swift.ir.IrVisibilityOpen
import com.voxeldev.mcodegen.dsl.scenario.ScenarioScope
import io.outfoxx.swiftpoet.Modifier
import io.outfoxx.swiftpoet.PropertySpec
import io.outfoxx.swiftpoet.TypeSpec
import org.jetbrains.kotlin.com.intellij.psi.PsiModifier
import org.jetbrains.kotlin.lexer.KtTokens

context(SwiftModule, ScenarioScope)
internal fun convertFields(
    irFields: List<IrField>,
    poetClassBuilder: TypeSpec.Builder,
    isTopLevel: Boolean = false,
) {
    irFields.forEach { irField ->
        poetClassBuilder.addProperty(
            convertField(irField, isTopLevel)
        )
    }
}
context(SwiftModule, ScenarioScope)
internal fun convertField(
    irField: IrField,
    isTopLevel: Boolean,
): PropertySpec {
    val fieldModifiers = getModifiers(irField, isTopLevel)

    return PropertySpec.builder(irField.name, convertType(irField.type)).apply {
        addModifiers(*fieldModifiers)

        mutable(irField.isMutable)

        // TODO: think of annotations support in any way

        // TODO: support initializer
    }.build()
}

context(SwiftModule, ScenarioScope)
private fun getModifiers(
    irField: IrField,
    isTopLevel: Boolean,
): Array<Modifier> {
    return buildList {
        when(irField.visibility) {
            is IrVisibilityPrivate ->  {
                if (isTopLevel) {
                    add(Modifier.FILEPRIVATE)
                } else {
                    add(Modifier.PRIVATE)
                }
            }

            is IrVisibilityPublic -> {
                if (irField.languageProperties[KtTokens.OPEN_KEYWORD.value] == true) {
                    add(Modifier.OPEN)
                } else {
                    add(Modifier.PUBLIC)
                }
            }

            is IrVisibilityInternal -> add(Modifier.INTERNAL)

            is IrVisibilityOpen -> add(Modifier.OPEN)

            is IrVisibilityFileprivate -> add(Modifier.FILEPRIVATE)
        }

        if (irField.languageProperties[PsiModifier.STATIC] == true) {
            add(Modifier.STATIC)
        }

        if (irField.languageProperties[KtTokens.FINAL_KEYWORD.value] == true) {
            add(Modifier.FINAL)
        }

        // TODO: support swift-specific modifiers
    }.toTypedArray()
}

File: swift-companion/Sources/LanguageModuleSwift/Source/Parse/ConvertFields.swift
//
//  ConvertFields.swift
//  swift-companion
//
//  Created by nVoxel on 19.06.2025.
//

import SyntaxSparrow
import AnyCodable

extension SyntaxTree {
    func convertField(swiftVariable: Variable) -> IrField {
        let name = swiftVariable.name
        let type = convertType(type: swiftVariable.type)

        let visibility = if (swiftVariable.isPrivate) {
            swiftVisibilityPrivate()
        } else if (swiftVariable.isPublic) {
            swiftVisibilityPublic()
        } else if (swiftVariable.isInternal) {
            swiftVisibilityInternal()
        } else if (swiftVariable.isOpen) {
            swiftVisibilityOpen()
        } else {
            swiftVisibilityFileprivate()
        }
        
        var languageProperties: [String : AnyCodable] = [:]
        
        if (swiftVariable.containsModifierWithKeyword(.static)) {
            languageProperties["static"] = true
        }
        
        if (swiftVariable.isFinal) {
            languageProperties["final"] = true
        }
        
        // TODO: Swift-specific modifiers
        
        let isMutable = swiftVariable.keyword == "var"
        
        // TODO: think of annotations support in any way

        // TODO: support initializer
        
        return IrField(
            name: name,
            irType: type,
            visibility: visibility,
            isMutable: isMutable,
            languageProperties: languageProperties
        )
    }
}
